### DISCUSSION TITLE
An upstream that filters too-new versions

### DISCUSSION THREAD

**Discussion by osher** [verdaccio/verdaccio#3029] [Q&A]
The use case is a security requirement, not to allow versions of 3rd party packages that are too new, while corporate packages must always be installed with latest and greatest.

The reason is a defense against new versions with yet-unknown exploits, or packages whose ownership has been compromised and have been loaded with malicious code.

Could such a behavior be accomplished some kind of middleware / plugin for `verdaccio`?

Basically, IMHO, all it needs to do is filter out from `package.json` sent replies to `npm info` versions that are X days old or younger, having this behavior configured per scope. 
This config can be on `config.yaml:packages` section.

Here's the furthest I got:
https://github.com/verdaccio/verdaccio/blob/5.x/src/api/endpoint/api/package.ts#L56
I do see the `config` is provided to that closure.
Mmm. 
If to be concrete - then we can add a concrete property like `config.yaml:packages[scope].minDaysSincePublication` or `.noNewerThan` or something of that sort, and filter `package.time` and `package.versions` accordingly.

If to be generic - then we need something like a pluggable `onPacakgeInfo` mapper/transformer that is instantiated with config and is called from here, which replaces some default `(package) => package` mapper or something.
In order to fail-fast on problems it should probably be loaded as last step of config-reading or around plugin load stage before the server is bound.

Thoughts? Ideas? Leads? References to something that does it already?
Any help will be appreciated!

**Comment by juanpicado** [comment]
There was a filter plugin initiative but not much tested, have you seen the website ?

**Reply by osher** [reply]
mm. https://verdaccio.org/docs/dev-plugins/#filter-plugin  Interesting.
Looks quite on target.

I'll give it a try and update 👍

**Reply by osher** [reply]
mm. looks like the doc in the site and the unit-tests do not agree with each other, or the example on the site is not clear.

The from example in the site:
```
filters:
   storage-filter-blackwhitelist:
     filter_file: /path/to/file
```
The impression I got from it is that `filter_file` is the module file, and it gets a link to the entire config document.

From the unit test:
https://github.com/verdaccio/verdaccio/pull/1161/files/b9ffac5d1bdf6f4a043a035900c91abe446b4441#diff-5ebf17640e9d7853559afa3f5ee24ce8b17c0650ef514bf1a5e31856f83f0b65R45-R50
I get the impression that the `filters` collection is a map of a requireable module-name -> config that is passed to it, and it does NOT get access to the entire config.

I think that access to the `uplinks` and `packages` is so useful that I took it for granted...
If I dont get access to these collecitons - I'll have to repeat information, probably with yaml-anchors - which is a hacky gotcha...

BTW - the `test/unit/api` tests disappear from the repo at Jun 13, 2019... 😱

**Reply by osher** [reply]
mm. there's also a problem with dist-tags. 
We can filter `time` and `versions` synchronously - that's simple. 

However, dist-tags is by definition a collection of "moving targets".
When we detect a dist-tag that points to a too-new version - we need to find what it was pointing before.

I think the simplest way is to return the package.json of the latest version in the pkg that is not too young - for this we need access to the store.

I mean - we could perform http-requests - but accessing `localhost` is just naive. We need again information from the higher config doc, and it will still be ugly. Better to provide the filter some interface to the store or actually the model that encapsulates both stores and upstreams.

AFAIK, we don't use dist-tags. But I can't know or guarantee that. Plus - I rather post a complete solution.

Thoughts?

**Comment by ashishbijlani** [comment]
I'm working on providing a Verdaccio filter plugin that invokes https://github.com/ossillate-inc/packj to block malicious and other "risky" packages. ETA: next month. I will post here when ready.
